src/crimson/common/errorator.h

```diff
@@ -544,6 +544,39 @@ struct errorator {
     template <class Func>
     void then(Func&&) = delete;
 
+    template <class ErrorVisitorT>
+    auto handle_error(ErrorVisitorT&& errfunc) {
+      static_assert((... && std::is_invocable_v<ErrorVisitorT,
+                                                AllowedErrors>),
+                    "provided Error Visitor is not exhaustive");
+      using return_errorator_t = make_errorator_t<
+        errorator<>,
+        std::decay_t<std::invoke_result_t<ErrorVisitorT, AllowedErrors>>...>;
+      using futurator_t = \
+        typename return_errorator_t::template futurize<::seastar::future<ValuesT...>>;
+      return this->then_wrapped(
+        [ errfunc = std::forward<ErrorVisitorT>(errfunc)
+        ] (auto&& future) mutable [[gnu::always_inline]] noexcept {
+          if (__builtin_expect(future.failed(), false)) {
+            return _safe_then_handle_errors<futurator_t>(
+              std::move(future), std::forward<ErrorVisitorT>(errfunc));
+          } else {
+            return typename futurator_t::type{ std::move(future) };
+          }
+        });
+    }
+    template <class ErrorFuncOne,
+              class ErrorFuncTwo,
+              class... ErrorFuncTail>
+    auto handle_error(ErrorFuncOne&& error_func_one,
+                      ErrorFuncTwo&& error_func_two,
+                      ErrorFuncTail&&... error_func_tail) {
+      return this->handle_error(
+        composer(std::forward<ErrorFuncOne>(error_func_one),
+                 std::forward<ErrorFuncTwo>(error_func_two),
+                 std::forward<ErrorFuncTail>(error_func_tail)...));
+    }
```

Comment:
i think we can simplify it like
```diff
diff --git a/src/crimson/common/errorator.h b/src/crimson/common/errorator.h
index 7e7baf5d67..f1c6a6e157 100644
--- a/src/crimson/common/errorator.h
+++ b/src/crimson/common/errorator.h
@@ -565,15 +565,12 @@ private:
           }
         });
     }
-    template <class ErrorFuncOne,
-              class ErrorFuncTwo,
+    template <class ErrorFuncHead,
               class... ErrorFuncTail>
-    auto handle_error(ErrorFuncOne&& error_func_one,
-                      ErrorFuncTwo&& error_func_two,
+    auto handle_error(ErrorFuncHead&& error_func_head,
                       ErrorFuncTail&&... error_func_tail) {
       return this->handle_error(
-        composer(std::forward<ErrorFuncOne>(error_func_one),
-                 std::forward<ErrorFuncTwo>(error_func_two),
+        composer(std::forward<ErrorFuncHead>(error_func_head),
                  std::forward<ErrorFuncTail>(error_func_tail)...));
     }
```

Reply:
Thanks for pointing this out, @tchaikov! I sent [similar clean-up](https://github.com/ceph/ceph/pull/31856) for `safe_then()`.